sdk/eventhub/Azure.Messaging.EventHubs/design/event-processor{T}-proposal.md

```diff
@@ -0,0 +1,520 @@
+                claimedOwnership.Add(ownership);
+            }
+        }
+
+        return Task.FromResult((IEnumerable<PartitionOwnership>)claimedOwnership);
+    }
+
+    private string CreateKey(
+        string fullyQualifiedNamespace,
+        string eventHubName,
+        string consumerGroup) => $"{ fullyQualifiedNamespace }/{ eventHubName }/{ consumerGroup }";
+}
+```
+
+## API skeleton
+
+### `Azure.Messaging.EventHubs.Specialized`
+
+```csharp
+public abstract class EventProcessor<TPartition> where TPartition : EventProcessorPartition, new()
+{
+    public string FullyQualifiedNamespace { get; }
+    public string EventHubName { get; }
+    public string ConsumerGroup { get; }
+    public string Identifier { get; protected set; }
+    public bool IsRunning { get; protected set; }
+    
+    protected EventProcessor(
+        int eventBatchSize,
+        string consumerGroup, 
+        string connectionString, 
+        EventProcessorOptions options = default);
+        
+    protected EventProcessor(
+        int eventBatchSize,
+        string consumerGroup, 
+        string connectionString, 
+        string eventHubName, 
+        EventProcessorOptions options = default);
+    
+    protected EventProcessor(
+        int eventBatchSize,
+        string consumerGroup,  
+        string fullyQualifiedNamespace, 
+        string eventHubName, 
+        TokenCredential credential, 
+        EventProcessorOptions options = default);
+    
+    // Optional extension points (core)
+    public virtual Task StartProcessingAsync(CancellationToken cancellationToken = default);
+    public virtual void StartProcessing(CancellationToken cancellationToken = default);
+    public virtual Task StopProcessingAsync(CancellationToken cancellationToken = default);
+    public virtual void StopProcessing(CancellationToken cancellationToken = default);
+    
+    protected virtual Task OnInitializingPartitionAsync(TPartition partition, CancellationToken cancellationToken);
+    protected virtual Task OnPartitionProcessingStoppedAsync(TPartition partition, ProcessingStoppedReason reason, CancellationToken cancellationToken);
+    
+    // Required extension points (core)
+    protected abstract Task OnProcessingEventBatchAsync(IEnumerable<EventData> events, TPartition partition, CancellationToken cancellationToken);
+    protected abstract Task OnProcessingErrorAsync(Exception exception, TPartition partition, string operationDescription, CancellationToken cancellationToken);
+    
+    // Required extension points (storage operations)
+    protected abstract Task<IEnumerable<EventProcessorCheckpoint>> ListCheckpointsAsync(CancellationToken cancellationToken);
+    protected abstract Task<IEnumerable<PartitionOwnership>> ListOwnershipAsync(CancellationToken cancellationToken);
+    protected abstract Task<IEnumerable<PartitionOwnership>> ClaimOwnershipAsync(IEnumerable<PartitionOwnership> desiredOwnership, CancellationToken cancellationToken);
```

Comment:
UpdateCheckpointAsync should be required as well, shouldn't it?

Reply:
Nope;  we don't ever call that, only user-provided code does.